src/client/pydaos/torch/torch_shim.c

```diff
@@ -0,0 +1,887 @@
+/**
+ * (C) Copyright 2019-2024 Intel Corporation.
+ * (C) Copyright 2024 Google LLC
+ * (C) Copyright 2024 Enakta Labs Ltd
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+#include <Python.h>
+
+#include <fcntl.h>  /* S_ISDIR */
+#include <libgen.h> /* dirname() */
+
+#include <daos.h>
+#include <daos_errno.h>
+#include <daos_event.h>
+#include <daos_types.h>
+#include <daos_fs.h>
+
+#include <gurt/debug.h>
+#include <gurt/common.h>
+
+#define PY_SHIM_MAGIC_NUMBER (0x7A8B)
+
+struct dfs_handle {
+	int           flags;
+	dfs_t        *dfs;
+	d_iov_t       global;
+
+	daos_handle_t eq;
+	pid_t         eq_owner_pid;
+};
+
+/* Parse arguments and magic number.
+ * As well as returning NULL this sets the Python exception state as required
+ */
+#define RETURN_NULL_IF_FAILED_TO_PARSE(args, format, x...)                                         \
+	do {                                                                                       \
+		int magic;                                                                         \
+		if (!PyArg_ParseTuple(args, "i" format, &magic, x)) {                              \
+			D_DEBUG(DB_ANY, "Bad args passed to %s", __func__);                        \
+			return NULL;                                                               \
+		}                                                                                  \
+		if (magic != PY_SHIM_MAGIC_NUMBER) {                                               \
+			D_ERROR("MAGIC number does not match, expected %d got %d\n",               \
+				PY_SHIM_MAGIC_NUMBER, magic);                                      \
+			PyErr_Format(PyExc_TypeError,                                              \
+				     "Bad magic value in torch(%s), expected %d got %d", __func__, \
+				     PY_SHIM_MAGIC_NUMBER, magic);                                 \
+			return NULL;                                                               \
+		}                                                                                  \
+	} while (0)
+
+static void
+atfork_handler(void)
+{
+	int rc = daos_reinit();
+	if (rc) {
+		D_WARN("daos_reinit() failed in child process %s (rc=%d)", d_errstr(rc), rc);
+	}
+}
+
+static PyObject *
+__shim_handle__module_init(PyObject *self, PyObject *args)
+{
+	int rc = daos_init();
+	if (rc) {
+		PyErr_Format(PyExc_TypeError, "Could not initialize DAOS module %s (rc=%d)",
+			     d_errstr(rc), rc);
+		return NULL;
+	}
+
+	rc = pthread_atfork(NULL, NULL, &atfork_handler);
+	if (rc) {
+		PyErr_Format(PyExc_TypeError, "Could not set atfork handler %s (rc=%d)",
+			     strerror(rc), rc);
+		return NULL;
+	}
+
+	return PyLong_FromLong(rc);
+}
+
+static PyObject *
+__shim_handle__module_fini(PyObject *self, PyObject *args)
+{
+	int rc = daos_fini();
+	if (rc == -DER_BUSY) {
+		/* Most likely module is shared via python multiprocessing.
+		 * The last one will free the resources */
+		rc = DER_SUCCESS;
+	}
+	return PyLong_FromLong(rc);
```

Comment:
dfs returns errno errors while daos api returns DER errors. Maybe it would be great for this shim layer to standardize on errno and not return DER errors given that most functions are the result of dfs calls? There is a function in libdaos to convert DER to errno.

Reply:
done